File: apps/embedding-service/priority_scheduler.py
from __future__ import annotations

from collections.abc import Iterator
from contextlib import contextmanager
from dataclasses import dataclass
from threading import Condition

INTERACTIVE_PRIORITY = "interactive"
BACKGROUND_PRIORITY = "background"

_PRIORITY_RANK = {
    INTERACTIVE_PRIORITY: 0,
    "question": 0,
    "foreground": 0,
    BACKGROUND_PRIORITY: 10,
}


def normalize_embedding_priority(value: str | None) -> str:
    normalized = (value or INTERACTIVE_PRIORITY).strip().lower()
    if _PRIORITY_RANK.get(normalized, 0) == 0:
        return INTERACTIVE_PRIORITY
    return BACKGROUND_PRIORITY


@dataclass(frozen=True)
class SchedulerSnapshot:
    active: bool
    waiting_interactive: int
    waiting_background: int


class EmbeddingPriorityScheduler:
    def __init__(self) -> None:
        self._condition = Condition()
        self._active = False
        self._next_sequence = 0
        self._waiting: list[tuple[int, int, str]] = []

    @contextmanager
    def slot(self, priority: str | None) -> Iterator[None]:
        normalized = normalize_embedding_priority(priority)
        rank = _PRIORITY_RANK[normalized]
        with self._condition:
            sequence = self._next_sequence
            self._next_sequence += 1
            entry = (rank, sequence, normalized)
            self._waiting.append(entry)

            while self._active or entry != min(self._waiting):
                self._condition.wait()

            self._waiting.remove(entry)
            self._active = True

        try:
            yield
        finally:
            with self._condition:
                self._active = False
                self._condition.notify_all()

    def snapshot(self) -> SchedulerSnapshot:
        with self._condition:
            return SchedulerSnapshot(
                active=self._active,
                waiting_interactive=sum(1 for rank, _, _ in self._waiting if rank == 0),
                waiting_background=sum(1 for rank, _, _ in self._waiting if rank > 0),
            )

File: apps/embedding-service/src/priority-scheduler.ts
export const INTERACTIVE_PRIORITY = "interactive";
export const BACKGROUND_PRIORITY = "background";

export type EmbeddingPriority =
  | typeof INTERACTIVE_PRIORITY
  | typeof BACKGROUND_PRIORITY;

const priorityRank = new Map<string, number>([
  [INTERACTIVE_PRIORITY, 0],
  ["question", 0],
  ["foreground", 0],
  [BACKGROUND_PRIORITY, 10]
]);

export interface SchedulerSnapshot {
  active: boolean;
  waiting_interactive: number;
  waiting_background: number;
}

interface Waiter {
  rank: number;
  sequence: number;
  normalized: EmbeddingPriority;
  resolve: () => void;
}

export const normalizeEmbeddingPriority = (
  value: string | null | undefined
): EmbeddingPriority => {
  const normalized = (value ?? INTERACTIVE_PRIORITY).trim().toLowerCase();
  return priorityRank.get(normalized) === 10
    ? BACKGROUND_PRIORITY
    : INTERACTIVE_PRIORITY;
};

export class EmbeddingPriorityScheduler {
  private active = false;
  private nextSequence = 0;
  private waiting: Waiter[] = [];

  async slot<T>(
    priority: string | null | undefined,
    task: () => Promise<T> | T
  ): Promise<T> {
    const normalized = normalizeEmbeddingPriority(priority);
    const waiter: Waiter = {
      rank: priorityRank.get(normalized) ?? 0,
      sequence: this.nextSequence,
      normalized,
      resolve: () => undefined
    };
    this.nextSequence += 1;
    await new Promise<void>((resolve) => {
      waiter.resolve = resolve;
      this.waiting.push(waiter);
      this.drain();
    });
    try {
      return await task();
    } finally {
      this.active = false;
      this.drain();
    }
  }

  snapshot(): SchedulerSnapshot {
    return {
      active: this.active,
      waiting_interactive: this.waiting.filter((waiter) => waiter.rank === 0)
        .length,
      waiting_background: this.waiting.filter((waiter) => waiter.rank > 0)
        .length
    };
  }

  private drain(): void {
    if (this.active || this.waiting.length === 0) {
      return;
    }
    const next = this.waiting.reduce((best, current) => {
      if (current.rank < best.rank) return current;
      if (current.rank === best.rank && current.sequence < best.sequence) {
        return current;
      }
      return best;
    });
    this.waiting = this.waiting.filter((waiter) => waiter !== next);
    this.active = true;
    queueMicrotask(next.resolve);
  }
}
